### DISCUSSION TITLE
Why, with openssl-3.2 and newer versions a simple self-signed certificate does not work as CA

### DISCUSSION THREAD

**Discussion by rababu1978** [openssl/openssl#26144] [Q&A] [answered]
self signed cert verification with openssl 3.1.3 works fine, but fails with 3.4.0. 

I did not see anything related to this in the release notes.

commands I use to create & verify the self signed
```
set OPENSSL_CONF=D:\openssl.cfg


openssl genrsa -des3 -out nfa-ca-key.pem -passout pass:raghucert 2048
openssl req -new -passin pass:raghucert -key nfa-ca-key.pem -out nfa-ca.csr -subj "/C=US/ST=Colorado/L=Colorado/O=Digicert Technologies/OU=Digicert/CN=Raghavendra Babu K/emailAddress=digicert@test.com"
copy nfa-ca-key.pem nfa-ca-key.pem.orig
openssl rsa -in nfa-ca-key.pem.orig -out nfa-ca-key.pem -passin pass:raghucert 
openssl x509 -passin pass:raghucert -req -days 1825 -in nfa-ca.csr -signkey nfa-ca-key.pem -out nfa-ca-cert.pem


openssl genrsa -des3 -out nfa-console-key.pem -passout pass:raghucert 2048
openssl req -new -passin pass:raghucert -key nfa-console-key.pem -out nfa-console.csr -subj "/C=US/ST=NewYork/L=NewYork/O=Digicert Inc./OU=AOD/CN=10.25.20.28/emailAddress=raghu.k@youtube.com"
copy nfa-console-key.pem nfa-console-key.pem.orig
openssl rsa -in nfa-console-key.pem.orig -out nfa-console-key.pem -passin pass:raghucert 
openssl x509 -passin pass:raghucert -req -extfile SAN.cnf -extensions req_ext -days 1825 -in nfa-console.csr -CA nfa-ca-cert.pem -CAkey nfa-ca-key.pem -set_serial 1111 -out nfa-console-cert.pem 

openssl verify -CAfile nfa-ca-cert.pem nfa-console-cert.pem
```
Above commands work in 3.1.3 not in 3.4.0

Any change went into 3.4.0 that I missed in the above commands? Appreciate your help!

**Comment by t8m** [comment, marked as answer]
Your _CA certificate_  as produced with the first x509 command is not really a CA certificate. With version 3.1 and older it is a version 1 certificate without extensions. However with later OpenSSL versions (3.2 and above) the produced certificate will be an end entity self signed version 3 certificate. 

This is mentioned in the NEWS.md for the 3.2 release: The `x509`, `ca`, and `req` apps now always produce X.509v3 certificates.

To produce a proper CA certificate in one step (no need to use separately `req` and `x509` commands) use:

` openssl req -new -x509 -passin pass:raghucert -key nfa-ca-key.pem -out nfa-ca-cert.pem -subj "/C=US/ST=Colorado/L=Colorado/O=Digicert Technologies/OU=Digicert/CN=Raghavendra Babu K/emailAddress=digicert@test.com" -addext "basicConstraints=critical,CA:true" -addext "keyUsage=critical,keyCertSign,cRLSign"`

**Comment by rababu1978** [comment]
Hi Tomáš Mráz,

Really appreciate for the quick response. I have applied the changes suggested but I get the same error. I did not combine req and x509 commands, sorry for that. Below are the step by step commands I used to verify the CA (not really) self-signed cert and the console cert signed by CA self signed cert.

C:\openssl\bin>set OPENSSL_CONF=C:\CA\NFA\Tools\openssl\bin\openssl.cfg

Generate the private key:
C:\openssl\bin>openssl genrsa -des3 -out nfa-ca-key.pem -passout pass:raghucert 2048

Generate a Certificate Signing Request:
C:\openssl\bin>openssl req -new -passin pass:raghucert -key nfa-ca-key.pem -out nfa-ca.csr -subj "/C=US/ST=Colorado/L=Colorado/O=Digicert Technologies/OU=Digicert/CN=Raghavendra Babu K/emailAddress=digicert@test.com" -addext "basicConstraints=critical,CA:true" -addext "keyUsage=critical,keyCertSign,cRLSign"

Remove passphrase from the private key:
C:\openssl\bin>copy nfa-ca-key.pem nfa-ca-key.pem.orig
        1 file(s) copied.

C:\openssl\bin>openssl rsa -in nfa-ca-key.pem.orig -out nfa-ca-key.pem -passin pass:raghucert
writing RSA key

Generate the simple self-signed certificate:
C:\openssl\bin>openssl x509 -req -days 1825 -in nfa-ca.csr -signkey nfa-ca-key.pem -out nfa-ca-cert.pem
Certificate request self-signature ok
subject=C=US, ST=Colorado, L=Colorado, O=Digicert Technologies, OU=Digicert, CN=Raghavendra Babu K, emailAddress=digicert@test.com


Below steps are to generate a new certificate for the Console, signed by the above CA certificate (not really CA) generated.

Generate the private key:
C:\openssl\bin>openssl genrsa -des3 -out nfa-console-key.pem -passout pass:raghucert 2048

Generate a Certificate Signing Request for console: 
C:\openssl\bin>openssl req -new -passin pass:raghucert -key nfa-console-key.pem -out nfa-console.csr -subj "/C=US/ST=Colorado/L=Colorado/O=Digicert Inc./OU=AOD/CN=10.20.22.17/emailAddress=raghu.k@youtube.com" -addext "basicConstraints=critical,CA:true" -addext "keyUsage=critical,keyCertSign,cRLSign"

Remove the passphrase from the private key:
C:\openssl\bin>copy nfa-console-key.pem nfa-console-key.pem.orig
        1 file(s) copied.

C:\openssl\bin>openssl rsa -in nfa-console-key.pem.orig -out nfa-console-key.pem -passin pass:raghucert
writing RSA key

Generate a certificate signed by the Console acting as a CA
C:\openssl\bin>openssl x509 -passin pass:raghucert -req -extfile SAN.cnf -extensions req_ext -days 1825 -in nfa-console.csr -CA nfa-ca-cert.pem -CAkey nfa-ca-key.pem -set_serial 1111 -out nfa-console-cert.pem
Certificate request self-signature ok
subject=C=US, ST=Colorado, L=Colorado, O=Digicert Inc., OU=AOD, CN=10.20.22.17, emailAddress=raghu.k@youtube.com


Verify that the certificates created are without errors.
C:\openssl\bin>openssl verify -CAfile nfa-ca-cert.pem nfa-console-cert.pem
C=US, ST=Colorado, L=Colorado, O=Digicert Technologies, OU=Digicert, CN=Raghavendra Babu K, emailAddress=digicert@test.com
error 79 at 1 depth lookup: invalid CA certificate
error nfa-console-cert.pem: verification failed

**Reply by t8m** [reply]
If you do not combine the req and x509 commands when generating the CA certificate you have to use ` -copy_extensions copy` option with the x509 command as the extensions won't be copied into the resulting certificate otherwise.

**Comment by rababu1978** [comment]
Hi Tomáš Mráz,
It worked with -copy_extensions copy/copyall.
Appreciate your help! Thanks.
